File: packages/guards/src/agent-budget-limit.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface AgentBudgetLimitOptions {
  action: 'block' | 'warn';
  /** Maximum budget in USD */
  maxBudget?: number;
  /** Cost added per check call */
  costPerCall?: number;
}

const COST_PATTERNS = [/\$[\d.]+/g, /cost:\s*[\d.]+/gi, /price:\s*[\d.]+/gi, /billing/gi];

export function agentBudgetLimit(options: AgentBudgetLimitOptions): Guard {
  const maxBudget = options.maxBudget ?? 10.0;
  const costPerCall = options.costPerCall ?? 0.01;
  let totalCost = 0;

  return {
    name: 'agent-budget-limit',
    version: '0.1.0',
    description: 'Tracks cumulative cost and enforces budget caps',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      totalCost += costPerCall;

      let costMentions = 0;
      for (const pattern of COST_PATTERNS) {
        const matches = text.match(pattern);
        if (matches) costMentions += matches.length;
      }

      const triggered = totalCost > maxBudget;
      return {
        guardName: 'agent-budget-limit',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { totalCost: Math.round(totalCost * 100) / 100, maxBudget, costMentions }
          : { totalCost: Math.round(totalCost * 100) / 100, maxBudget, costMentions },
      };
    },
  };
}

File: python/open_guardrail/guards/agent_budget_limit.py
"""Tracks cumulative cost and enforces budget caps."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_COST_PATTERNS = [
    re.compile(r"\$[\d.]+"),
    re.compile(r"cost:\s*[\d.]+", re.IGNORECASE),
    re.compile(r"price:\s*[\d.]+", re.IGNORECASE),
    re.compile(r"billing", re.IGNORECASE),
]


class _AgentBudgetLimit:
    def __init__(
        self,
        *,
        action: str = "block",
        max_budget: float = 10.0,
        cost_per_call: float = 0.01,
    ) -> None:
        self.name = "agent-budget-limit"
        self.action = action
        self.max_budget = max_budget
        self.cost_per_call = cost_per_call
        self._total_cost: float = 0.0

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        self._total_cost += self.cost_per_call

        cost_mentions = 0
        for pattern in _COST_PATTERNS:
            cost_mentions += len(pattern.findall(text))

        triggered = self._total_cost > self.max_budget
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="agent-budget-limit",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Budget exceeded: ${self._total_cost:.2f} / ${self.max_budget:.2f}" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"total_cost": round(self._total_cost, 2), "max_budget": self.max_budget, "cost_mentions": cost_mentions},
        )


def agent_budget_limit(
    *,
    action: str = "block",
    max_budget: float = 10.0,
    cost_per_call: float = 0.01,
) -> _AgentBudgetLimit:
    return _AgentBudgetLimit(
        action=action,
        max_budget=max_budget,
        cost_per_call=cost_per_call,
    )
